### PR TITLE
Handle non-existing file in Backup password check too

### PR DESCRIPTION
**Pull request by agners**
## Proposed change
Make sure we handle a non-existing backup file also when validating the password.

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (which adds functionality to the supervisor)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:
- Link to cli pull request:
- Link to client library pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Ruff (`ruff format supervisor tests`)
- [ ] Tests have been added to verify that the new code works.

If API endpoints or add-on configuration are added/changed:

- [ ] Documentation added/updated for [developers.home-assistant.io][docs-repository]
- [ ] [CLI][cli-repository] updated (if necessary)
- [ ] [Client library][client-library-repository] updated (if necessary)

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[docs-repository]: https://github.com/home-assistant/developers.home-assistant
[cli-repository]: https://github.com/home-assistant/cli
[client-library-repository]: https://github.com/home-assistant-libs/python-supervisor-client/

## Summary by CodeRabbit

- **Bug Fixes**
	- Enhanced error handling for backup operations to present clearer messages when backups are missing or cannot be read.
	- Improved the backup password validation process by enabling a smoother recovery path when specific errors occur.
	- Streamlined logic for validating backup locations and passwords to ensure accuracy based on specified locations.

- **New Features**
	- Added a new test function to validate the password for backups, covering various scenarios and edge cases.

- **Tests**
	- Updated test functions to allow dynamic selection of backup files, improving flexibility and robustness of test scenarios.
	- Refined handling of backup instance properties during tests, enhancing clarity and maintainability.

### CODE REVIEW

File: supervisor/backups/manager.py [hunk 1]
```diff
@@ -727,10 +727,14 @@ async def _validate_location_password(
...
-                raise BackupInvalidError(
-                    f"Invalid password for backup {backup.slug}", _LOGGER.error
-                )
+            try:
+                if not await backup.validate_password():
+                    raise BackupInvalidError(
+                        f"Invalid password for backup {backup.slug}", _LOGGER.error
+                    )
+            except BackupFileNotFoundError:
+                self.sys_create_task(self.sys_backups.reload(location))
```
Comment (mdegat01) [inline supervisor/backups/manager.py:736]:
```suggestion
                self.sys_create_task(self.reload(location))
```
`sys_backups` is how other objects get to backup manager but that's where we are already. Although we could just do this call in the other `except` before raising `BackupFileNotFoundError` and drop this `try` `except` if we prefer, we don't have to schedule the `reload` from here specifically.

**Review by mdegat01 — CHANGES_REQUESTED** [review]
Should probably have a test case as well. Should just be able to add a parameter set of a partial restore with an invalid password to here:
https://github.com/home-assistant/supervisor/blob/4c04f364a3bd265842bbc765fed7a7c2e7e00fa2/tests/api/test_backups.py#L1122-L1163

File: supervisor/backups/manager.py (hunk 1 above)
Comment (agners) [reply supervisor/backups/manager.py:736]: We don't have the location in the other try except...

Comment (mdegat01) [reply supervisor/backups/manager.py]: oh, good point  👍

**Comment by mdegat01** [comment]
@agners in retrospect, our conversation above should've raised warning flags to me. The `validate_password` method needs to know the location, otherwise it cannot check the password correctly. Since we now support having the same backup in mulitiple locations with different encryption options, we cannot validate the password without knowing the location. We were only validating the password using the default location even if that's not where we were restoring from.

I adjusted the code to fix this bug, removed the unnecessary `try` `catch` and added a test case to keep this all working going forward.

**Review by mdegat01 — APPROVED** [review]

**Comment by agners** [comment]
> @agners in retrospect, our conversation above should've raised warning flags to me. The `validate_password` method needs to know the location, otherwise it cannot check the password correctly. Since we now support having the same backup in mulitiple locations with different encryption options, we cannot validate the password without knowing the location. We were only validating the password using the default location even if that's not where we were restoring from.

Uh interesting, that could explain some of the issues we have been seeing that the password was not considered valid in some cases maybe :thinking: 


> I adjusted the code to fix this bug, removed the unnecessary `try` `catch` and added a test case to keep this all working going forward.

Cool, thanks! There was some more fallout in not properly mocked test cases (the test cases mocked the `protected` property directly, instead of the new ultimate source of truth in `all_locations`. Tests should be fine now.

File: supervisor/backups/backup.py [hunk 2]
```diff
@@ -433,7 +433,13 @@ def _validate_file() -> bool:
                     _LOGGER.exception("Unexpected error validating password")
                     return True
 
-        return await self.sys_run_in_executor(_validate_file)
+        try:
+            await self.sys_run_in_executor(_validate_file)
```
Comment (agners) [inline supervisor/backups/backup.py]: The return statement here was actually missing :see_no_evil: 

I've added a test to cover `validate_file`, with that this should no longer happen.
